### DISCUSSION TITLE
Question about plotting pareto front

### DISCUSSION THREAD

**Discussion by YouHyin** [meta-pytorch/botorch#1666] [Q&A] [answered]
I have a question about this code.

```
from matplotlib.cm import ScalarMappable

fig, axes = plt.subplots(1, 4, figsize=(23, 7), sharex=True, sharey=True)
algos = ["Sobol", "qNParEGO", "qEHVI", "qNEHVI"]
cm = plt.cm.get_cmap('viridis')

** ### batch_number = torch.cat(
    [torch.zeros(2*(problem.dim+1)), torch.arange(1, N_BATCH+1).repeat(BATCH_SIZE, 1).t().reshape(-1)]
).numpy()**
for i, train_obj in enumerate((train_obj_true_random, train_obj_true_qparego, train_obj_true_qehvi, train_obj_true_qnehvi)):
    sc = axes[i].scatter(
        train_obj[:, 0].cpu().numpy(), train_obj[:,1].cpu().numpy(), c=batch_number, alpha=0.8,
    )
    axes[i].set_title(algos[i])
    axes[i].set_xlabel("Objective 1")
axes[0].set_ylabel("Objective 2")
norm = plt.Normalize(batch_number.min(), batch_number.max())
sm =  ScalarMappable(norm=norm, cmap=cm)
sm.set_array([])
fig.subplots_adjust(right=0.9)
cbar_ax = fig.add_axes([0.93, 0.15, 0.01, 0.7])
cbar = fig.colorbar(sm, cax=cbar_ax)
cbar.ax.set_title("Iteration")
```

What role does batch_number play when drawing the Pareto front?
I don't understand the code of batch_number I marked.
I would really appreciate it if you could explain it.

Thank you!

**Comment by Balandat** [comment, marked as answer]
`batch_number` here is just an index for the batch of points that was generated during the BO Loop (in a sequential setting with `BATCH_SIZE=1`) this is identical to the trial index (modulo the first batch, which is a random exploration batch of size `2*(1+problem_dim)`). Here this is used to color the points in the Pareto Front visualization so we can see how the optimization progresses in the outcome space.

**Reply by YouHyin** [reply]
I'm so thankful. Thanks to you, I understand perfectly!
